src/vm/virtualMachine.ts

```diff
@@ -150,224 +108,192 @@ export class VirtualMachine {
   // fetch them from memory and if it fails to do so, it can
   // deduce them from the instruction itself.
   computeOperands(instruction: Instruction): Operands {
-    let res: SegmentValue | undefined = undefined;
-
     const {
       dstOffset,
       op0Offset,
       op1Offset,
       dstRegister,
       op0Register,
       op1Source,
+      resLogic,
+      opcode,
     } = instruction;
 
-    // Compute the destination address based on the dstRegister and
-    // the offset
-    // Example: const dstAddr = this.fp.add(5) == fp + 5;
-    // const dstAddr = this[instruction.dstRegister].add(instruction.dstOffset);
-    const dstAddr =
-      dstRegister === Register.Fp
-        ? this.fp.add(dstOffset)
-        : this.ap.add(dstOffset);
-    let dst = this.memory.get(dstAddr);
-
-    // Compute the first operand address based on the op0Register and
-    // the offset
-    // const op0Addr = this[instruction.op0Register].add(instruction.op0Offset);
-    const op0Addr =
-      op0Register === Register.Fp
-        ? this.fp.add(op0Offset)
-        : this.ap.add(op0Offset);
-
-    let op0 = this.memory.get(op0Addr);
-
-    // Compute the second operand address based on the op1Source and
-    // the offset
-    const op1Addr = this.computeOp1Address(op1Source, op1Offset, op0);
-    let op1 = this.memory.get(op1Addr);
-
-    // If op0 is undefined, then we can deduce it from the instruction, dst and op1
-    // We also deduce the result based on the result logic
-    if (op0 === undefined) {
-      const deducedOp0 = this.deduceOp0(instruction, dst, op1);
-
-      if (deducedOp0 !== undefined) {
-        this.memory.assertEq(op0Addr, deducedOp0);
-      }
-      op0 = deducedOp0;
-    }
+    let dstAddr: Relocatable;
+    let op0Addr: Relocatable;
+    let op1Addr: Relocatable;
 
-    // If operand 1 is undefined, then we can deduce it from the instruction,
-    // destination and operand 0.
-    // We also deduce the result based on the result logic
-    if (op1 === undefined) {
-      const deducedOp1 = this.deduceOp1(instruction, dst, op0);
+    let res: SegmentValue | undefined;
+    let dst: SegmentValue | undefined;
+    let op0: SegmentValue | undefined;
+    let op1: SegmentValue | undefined;
 
-      if (deducedOp1 !== undefined) {
-        this.memory.assertEq(op1Addr, deducedOp1);
-      }
+    // Compute the destination address based
+    // on the dstRegister and the offset
+    // Example: const dstAddr = this.fp.add(5) == fp + 5;
+    switch (dstRegister) {
```

Comment:
```
dstAddr = this.registers[dstRegister].add(dstOffset)
// etc.
```

and throw errors while decoding the instruction

Reply:
To do so, we need to define a mapping from the Register enum values to the corresponding attribute. 
As it is only used in the `computeOperands()` method (renamed `computeAuxValues()` in more recent commits), defining it as a variable makes more sense than defining an attribute imo

```typescript
computeAuxValues(instruction: Instruction) {
  // [...]

  const registers = {
    [Register.Pc]: this.pc,
    [Register.Ap]: this.ap,
    [Register.Fp]: this.fp,
  };

  const dstAddr = registers[dstRegister].add(dstOffset);
  
// [...]
}
```

The errors thrown in the default switch cases were actually unnecessary as already handled when decoding instructions

Reply:
what about removing the `this.pc`, `this.ap` and `this.pc` and only have `this.registers`?